File: queue-stack/LinkedListQueueAlg.java
public class LinkedListQueueAlg {
  public static void main(String[] args) {
    LinkedListQueue lq = new LinkedListQueue();
    lq.enqueue("a");
    lq.enqueue("b");
    lq.enqueue("c");
    System.out.println(lq.dequeue());

    lq.enqueue("d");
    System.out.println(lq.dequeue());
    
    System.out.println(lq.dequeue());
    System.out.println(lq.dequeue());
  }
}

class LinkedListQueue {
  private Node first, last;

  private class Node {
    String item;
    Node next;

    public String toString() {
      return item;
    }
  }

  public boolean isEmpty() {
    return first == null;
  }

  public void enqueue(String item) {
    Node previousLast = last;
    last = new Node();
    last.item = item;
    last.next = null;

    // Handle epmty queue
    if (isEmpty()) {
      first = last;
    } else {
      previousLast.next = last;
    }
  }

  public String dequeue() {
    String item = first.item;
    first = first.next;
    // Handle epmty queue
    if (isEmpty()) last = null;
    
    return item;
  }
}

File: queue-stack/LinkedListQueueAlg.ts
class LinkedListQueueAlg {
  public static main(): void {
    var lq: LinkedListQueue = new LinkedListQueue();
    lq.enqueue("a");
    lq.enqueue("b");
    lq.enqueue("c");
    console.log(lq.dequeue());

    lq.enqueue("d");
    console.log(lq.dequeue());
    
    console.log(lq.dequeue());
    console.log(lq.dequeue());
  }
}

class _Node {
  public item: string;
  public next: _Node;

  public toString(): string {
    return this.item;
  }
}

class LinkedListQueue {
  private first: _Node = null;
  private last: _Node = null;

  public isEmpty(): boolean {
    return this.first == null;
  }

  public enqueue(item: string): void {
    var previousLast: _Node = this.last;
    this.last = new _Node();
    this.last.item = item;
    this.last.next = null;

    if (this.isEmpty()) {
      this.first = this.last;
    } else {
      previousLast.next = this.last;
    }
  }

  public dequeue(): string {
    var item: string = this.first.item;
    this.first = this.first.next;
    if (this.isEmpty()) {
      this.last = null;
    }

    return item;
  }
}

LinkedListQueueAlg.main();